features/netsocket/NetworkInterface.cpp

```diff
@@ -79,6 +81,60 @@ void NetworkInterface::attach(mbed::Callback<void(nsapi_event_t, intptr_t)> stat
 {
 }
 
+
+typedef struct iface_eventlist_entry
+{
+    NetworkInterface *iface;
+    void (*status_cb)(nsapi_event_t, intptr_t);
+    ns_list_link_t link;
+} iface_eventlist_entry_t;
+
+typedef NS_LIST_HEAD(iface_eventlist_entry_t, link) iface_eventlist_t;
+
+static iface_eventlist_t *get_interface_event_list_head()
+{
+    static iface_eventlist_t NS_LIST_NAME_INIT(event_list);
+    return &event_list;
+}
+
+static void call_all_event_listeners(NetworkInterface *iface, nsapi_event_t event, intptr_t val)
+{
+    iface_eventlist_t *event_list = get_interface_event_list_head();
+    ns_list_foreach(iface_eventlist_entry_t, entry, event_list) {
+        if (entry->iface == iface) {
+            entry->status_cb(event, val);
+        }
+    }
+}
+
+static void modify_event_listeners(NetworkInterface *iface, void (*status_cb)(nsapi_event_t, intptr_t), bool remove)
+{
+    iface_eventlist_t *event_list = get_interface_event_list_head();
+    if (remove) {
+        ns_list_foreach_safe(iface_eventlist_entry_t, entry, event_list) {
+            if (entry->status_cb == status_cb && entry->iface == iface) {
+                ns_list_remove(event_list, entry);
+                return;
+            }
+        }
+    } else {
+        iface_eventlist_entry_t *entry = new iface_eventlist_entry_t;
+        entry->iface = iface;
+        entry->status_cb = status_cb;
+        ns_list_add_to_end(event_list, entry);
+        iface->attach(mbed::callback(&call_all_event_listeners, iface));
+    }
+}
+
+void NetworkInterface::add_event_listener(void (*status_cb)(nsapi_event_t, intptr_t))
```

Comment:
How about if you give NULL here as status_cb?

Reply:
Why would you?

It is not acceptable value and leads to HardFault.

Reply:
Being protective on your API and checking for NULL values can seriously hide bugs where your data types or pointers gets corrupted and are overwritten with zeros.
Then when that data is fed to here, it will be silently ignored, and device keeps running, but actually never receives callbacks.

Opposite would be that data corruption would immediately lead to HardFault when fed into APIs that try to access the value.

These Hardfault are quickly catched in the testing whereas silently ignoring carbage might lead to very hard to debug issues, as it might take long time to capture your failure and there is no knowledge anymore where it went wrong.

You should only do protective programming in external interfaces where you can expect carbage.